### PR TITLE
Async Client: Asynchronous Queries for Async Client

### PR DESCRIPTION
**Pull request by alkatrivedi**
This PR contains the hand written code layer for Asynchronous implementation of the Async Client Queries.

### CODE REVIEW

File: samples/samples/async_client_sample.py [hunk 1]
```diff
@@ -0,0 +1,648 @@
...
+# limitations under the License.
+#
+
+"""
+This application demonstrates how to do basic asynchronous operations using Cloud Spanner.
+"""
+
+from asyncore import readwrite
+from curses import meta
+from email.message import Message
```
Comment (rajatbhatta) [inline samples/samples/async_client_sample.py:23]: Remove all unused imports.

Comment (rajatbhatta) [inline samples/samples/async_client_sample.py:23]: Remove all unused imports.

File: samples/samples/async_client_sample.py [hunk 2]
```diff
@@ -0,0 +1,648 @@
...
+    # Creates the instance object associated with instance id
+    instance = spanner_client.instance(instance_id)
+
+    # Creates the database object associated with instance object
+    database_object = instance.database(database_id)
+
+    # Returns async client object and database object
+    return async_client, database_object
+
+# [START async client partition read]
```
Comment (rajatbhatta) [inline samples/samples/async_client_sample.py:70]: Use underscore as separator for tags.

File: samples/samples/async_client_sample.py [hunk 3]
```diff
@@ -0,0 +1,648 @@
...
+
+# [START async client rollback]
+async def rollback(database_object, async_client, task_num):
+
+     print("\nStarting task: ", task_num)
+
+     # Create the session object associated with the database object
+     session_object = session.Session(database_object)
+
+     """Create the session."""
```
Comment (rajatbhatta) [inline samples/samples/async_client_sample.py:145]: You can have comments starting with # for comments like this.

File: samples/samples/async_client_sample.py [hunk 4]
```diff
@@ -0,0 +1,648 @@
...
+    )
+    subparsers.add_parser(
+        "partition_read",
+        help=partition_read.__doc__
+    )
+    args = parser.parse_args()
+
+    async_client, database_object = getObjects(args.instance_id, args.database_id)
+
+    if args.command == "create_session":
```
Comment (rajatbhatta) [inline samples/samples/async_client_sample.py:565]: As all your conditions are of the type x == value with x being the same in every condition, switch case usually makes sense.

**Review by rajatbhatta — COMMENTED** [review]
Here are a few points I've from my first look. I'll take a look at the implementation in more detail and add more comments later if need be.

File: samples/samples/async_client_sample.py [hunk 5]
```diff
@@ -0,0 +1,648 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+This application demonstrates how to do basic asynchronous operations using Cloud Spanner.
+"""
+
+from asyncore import readwrite
```
Comment (rajatbhatta) [inline samples/samples/async_client_sample.py:21]: Have proper formatting for imports (grouping and ordering within groups). go/pystyle#imports-formatting

File: samples/samples/async_client_sample.py [hunk 6]
```diff
@@ -0,0 +1,648 @@
...
+
+     print("\nStarting task: ", task_num)
+
+     # Create the session object associated with the database object
+     session_object = session.Session(database_object)
+
+     """Create the session."""
+     session_object.create()
+     statements = ExecuteBatchDmlRequest.Statement(
+        sql = "INSERT INTO Singers (SingerId, FirstName, LastName) Values (1003, 'Alka', 'Trivedi')",
```
Comment (rajatbhatta) [inline samples/samples/async_client_sample.py:270]: Try sticking to Maximum line length of 80 characters.

File: samples/samples/async_client_sample.py [hunk 7]
```diff
@@ -0,0 +1,648 @@
...
+
+# [START async client batch_create_session]
+async def batch_create_session(database_object, async_client, task_num):
+
+     print("\nStarting task: ", task_num)
+
+     """Request object for batch create session."""
+     request = BatchCreateSessionsRequest(
+         database = database_object.name,
+         session_count=1420,
```
Comment (rajatbhatta) [inline samples/samples/async_client_sample.py:451]: How did we arrive at this session count here? Or are we setting this to a random value?

File: samples/samples/async_client_sample.py [hunk 8]
```diff
@@ -0,0 +1,648 @@
...
+                 ),
+             ),
+         ),
+         sql =  "UPDATE Singers SET FirstName = 'Trivedi' WHERE SingerId = 1003",
+     )
+
+     """Partition the sql statement query."""
+     response = await async_client.partition_query(request = request)
+
+     """Prints the response."""
```
Comment (rajatbhatta) [inline samples/samples/async_client_sample.py:131]: Try avoiding self-explanatory comments like this.

Quoting from: go/pystyle#comments

`Never describe the code. Assume the person reading the code knows Python (though not what you're trying to do) better than you do.`

File: samples/samples/async_client_sample.py [hunk 9]
```diff
@@ -0,0 +1,648 @@
...
+     """Prints the session got."""
+     print(response)
+
+     print("\nEnding task: ", task_num)
+
+# [END async client get_session]
+
+
+# [START async client batch_create_session]
+async def batch_create_session(database_object, async_client, task_num):
```
Comment (rajatbhatta) [inline samples/samples/async_client_sample.py:444]: batch_create_session -> batch_create_sessions

File: samples/samples/async_client_sample.py [hunk 10]
```diff
@@ -0,0 +1,648 @@
...
+     print("\nEnding task: ", task_num)
+
+# [END async client create_session]
+
+
+async def main():
+    parser = argparse.ArgumentParser(
+        description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter
+    )
+    parser.add_argument("--instance_id", help="Your Cloud Spanner instance ID.")
```
Comment (rajatbhatta) [inline samples/samples/async_client_sample.py:490]: Can we provide a default value for instance as well? (as we have provided it for database_id)

Or we could skip the default value for database_id as well.

File: samples/samples/async_client_sample.py [hunk 11]
```diff
@@ -0,0 +1,648 @@
...
+    parser.add_argument(
+        "--session_name", help="session name."
+    )
+    parser.add_argument(
+        "--n", help="number of queries."
+    )
+    subparsers = parser.add_subparsers(dest="command")
+    subparsers.add_parser(
+        "create_session",
+        help=create_session.__doc__
```
Comment (rajatbhatta) [inline samples/samples/async_client_sample.py:503]: Can we add docstrings to each of these functions so that <function_name>.__doc__ returns the method documentation?

File: samples/samples/async_client_sample.py [hunk 12]
```diff
@@ -0,0 +1,648 @@
...
+     """Prints the sessions batchwise."""
+     print(response)
+
+     print("\nEnding task: ", task_num)
+
+# [END async client batch_create_session]
+
+
+# [START async client create_session]
+async def create_session(database_object, async_client, task_num):
```
Comment (rajatbhatta) [inline samples/samples/async_client_sample.py:466]: task_num -> task_id?

File: samples/samples/async_client_sample.py [hunk 13]
```diff
@@ -0,0 +1,648 @@
...
+
+     """Create the session."""
+     session_object.create()
+
+     """Request object for read."""
+     request = ReadRequest(
+         session = session_object.name,
+         table =  'Singers',
+         columns = ["SingerId", "FirstName", "LastName"],
+         key_set = keyset.KeySet(all_=True)._to_pb(),
```
Comment (rajatbhatta) [inline samples/samples/async_client_sample.py:246]: Names with a leading underscore indicate to other programmers that the attribute or method is intended to be private (However, privacy is not enforced in any way). We should avoid calling _to_pb() from here, as it it intended to be private.

File: samples/samples/async_client_sample.py (hunk 4 above)
Comment (alkatrivedi) [reply samples/samples/async_client_sample.py:565]: As python does not have an in-built switch case statement should I add a class to implement that?

File: samples/samples/async_client_sample.py (hunk 13 above)
Comment (alkatrivedi) [reply samples/samples/async_client_sample.py:246]: 'key_set' was the required parameter for making this request. And this parameter requires calling _to_pb() method. Hence, I am not sure if there is any way we can avoid calling it.

File: samples/samples/async_client_sample.py (hunk 1 above)
Comment (alkatrivedi) [reply samples/samples/async_client_sample.py:23]: Removed.

File: samples/samples/async_client_sample.py (hunk 5 above)
Comment (alkatrivedi) [reply samples/samples/async_client_sample.py:21]: Done.

File: samples/samples/async_client_sample.py (hunk 6 above)
Comment (alkatrivedi) [reply samples/samples/async_client_sample.py:270]: Done.

Comment (rajatbhatta) [inline samples/samples/async_client_sample.py:270]: Try not to have own name in SQL here.

File: samples/samples/async_client_sample.py (hunk 7 above)
Comment (alkatrivedi) [reply samples/samples/async_client_sample.py:451]: We are setting it to a random value. This parameter is the number of sessions to be created in this batch call, hence we can set any value to it.

File: samples/samples/async_client_sample.py (hunk 8 above)
Comment (alkatrivedi) [reply samples/samples/async_client_sample.py:131]: Done.

File: samples/samples/async_client_sample.py (hunk 9 above)
Comment (alkatrivedi) [reply samples/samples/async_client_sample.py:444]: Done.

File: samples/samples/async_client_sample.py (hunk 10 above)
Comment (alkatrivedi) [reply samples/samples/async_client_sample.py:490]: Done

File: samples/samples/async_client_sample.py (hunk 11 above)
Comment (alkatrivedi) [reply samples/samples/async_client_sample.py:503]: Done

File: samples/samples/async_client_sample.py (hunk 12 above)
Comment (alkatrivedi) [reply samples/samples/async_client_sample.py:466]: Done

File: samples/samples/async_client_sample.py [hunk 14]
```diff
@@ -0,0 +1,648 @@
...
+     print("\nStarting task: ", task_num)
+
+     # Create the session object associated with the database object
+     session_object = session.Session(database_object)
+
+     """Create the session."""
+     session_object.create()
+
+     """Read write transaction object."""
+     transactionOption_object = TransactionOptions(
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply samples/samples/async_client_sample.py:210]: Done

File: samples/samples/async_client_sample.py (hunk 2 above)
Comment (alkatrivedi) [reply samples/samples/async_client_sample.py:70]: Done

File: samples/samples/async_client_sample.py (hunk 4 above)
Comment (rajatbhatta) [reply samples/samples/async_client_sample.py:565]: No, you need not implement it.

File: samples/samples/async_client_sample.py (hunk 6 above)
Comment (alkatrivedi) [reply samples/samples/async_client_sample.py:270]: Done

File: samples/samples/async_client_sample.py [hunk 15]
```diff
@@ -0,0 +1,637 @@
...
+                 ),
+             ),
+         ),
+         table = 'Singers',
+         key_set = keyset.KeySet(all_=True)._to_pb(),
+     )
+
+     response = await async_client.partition_read(request = request)
+
+     print(response)
```
Comment (asthamohta) [inline samples/samples/async_client_sample.py:93]: rather than print in all the fxns have return

File: samples/samples/async_client_sample.py [hunk 16]
```diff
@@ -0,0 +1,637 @@
...
+    return async_client, database_object
+
+
+# [START async_client_partition_read]
+async def partition_read(database_object, async_client, task_id):
+
+     """Creates a set of partition tokens that can be
+     used to execute a read operation in parallel."""
+
+     print("\nStarting task: ", task_id)
```
Comment (asthamohta) [inline samples/samples/async_client_sample.py:72]: remove the starting task from all fxns

File: samples/samples/async_client_sample.py [hunk 17]
```diff
@@ -0,0 +1,637 @@
...
+    instance = spanner_client.instance(instance_id)
+
+    database_object = instance.database(database_id)
+
+    return async_client, database_object
+
+
+# [START async_client_partition_read]
+async def partition_read(database_object, async_client, task_id):
+
```
Comment (asthamohta) [inline samples/samples/async_client_sample.py:68]: nit: remove all the extra line to make the code more compact

File: samples/samples/async_client_sample.py [hunk 18]
```diff
@@ -0,0 +1,637 @@
...
+import tracemalloc
+import argparse
+import time
+
+
+# [ FUNCTION for creating async client and spanner client object]
+def get_async_client_and_associate_database(instance_id, database_id):
+    
+    async_client = spanner.SpannerAsyncClient()
+
```
Comment (asthamohta) [inline samples/samples/async_client_sample.py:56]: nit: here as well remove all the extra lines

File: tests/system/_helpers.py [hunk 19]
```diff
@@ -1,134 +0,0 @@
-# Copyright 2021 Google LLC All rights reserved.
```
Comment (asthamohta) [inline tests/system/_helpers.py:1]: Why is this file deleted? Use checkout instead

File: samples/samples/async_client_sample.py (hunk 15 above)
Comment (alkatrivedi) [reply samples/samples/async_client_sample.py]: Done

File: samples/samples/async_client_sample.py (hunk 16 above)
Comment (alkatrivedi) [reply samples/samples/async_client_sample.py]: Done

File: samples/samples/async_client_sample.py (hunk 17 above)
Comment (alkatrivedi) [reply samples/samples/async_client_sample.py:63]: Done

File: samples/samples/async_client_sample.py (hunk 18 above)
Comment (alkatrivedi) [reply samples/samples/async_client_sample.py]: Done
